### PR TITLE
[13.0][ADD] base_ebill_payment_contract

### PR DESCRIPTION
**Pull request by TDu**
The goal of this module is to be a common ground for ebilling addons.
Alone this module has no purpose. It needs to be installed with a module
that implements a specific ebilling system.
This depending module has to implement his own 'transmit.method'.

### CODE REVIEW

**Comment by sebalix** [comment]
Depends on https://github.com/OCA/account-invoicing/pull/682

File: base_ebill_payment_contract/models/ebill_payment_contract.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import api, fields, models
+
+
+class EbillPaymentContract(models.Model):
+
+    _name = "ebill.payment.contract"
+    _description = "Ebill Payment Contract"
+    _rec_name = "name"
```
Comment (sebalix) [inline base_ebill_payment_contract/models/ebill_payment_contract.py]:
```suggestion
```
As it's already the default

File: base_ebill_payment_contract/models/ebill_payment_contract.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+    transmit_method_id = fields.Many2one(
+        comodel_name="transmit.method", string="Service Name", ondelete="restrict",
+    )
+    partner_id = fields.Many2one(
+        comodel_name="res.partner", required=True, string="Customer"
+    )
+    partner_id = fields.Many2one(
+        comodel_name="res.partner", required=True, string="Customer"
+    )
```
Comment (sebalix) [inline base_ebill_payment_contract/models/ebill_payment_contract.py]: Field `partner_id` declared twice

File: base_ebill_payment_contract/models/ebill_payment_contract.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        It is valid if the contract is opened and his start date is in the pass
+        And his end date is in the future or not set.
+        """
+        for contract in self:
+            contract.is_valid = (
+                contract.state == "open"
+                and contract.date_start
+                and contract.date_start <= fields.Date.today()
+                and (not contract.date_end or contract.date_end >= fields.Date.today())
+            )
```
Comment (sebalix) [inline base_ebill_payment_contract/models/ebill_payment_contract.py]:
```suggestion
        today = fields.Date.today()
        for contract in self:
            contract.is_valid = (
                contract.state == "open"
                and contract.date_start
                and contract.date_start <= today
                and (not contract.date_end or contract.date_end >= today)
            )
```

File: base_ebill_payment_contract/views/ebill_payment_contract.xml [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    <record model="ir.ui.view" id="ebill_payment_contract_search_view">
+        <field name="name">ebill.payment.contract.search</field>
+        <field name="model">ebill.payment.contract</field>
+        <field name="arch" type="xml">
+            <search>
+                <field name="transmit_method_id" />
+                <field name="partner_id" />
+                <filter
+                    name="active"
+                    string="Active"
```
Comment (sebalix) [inline base_ebill_payment_contract/views/ebill_payment_contract.xml]: As `Active` has a different sense in Odoo, better to rename this `is_valid / Valid`?

**Review by sebalix — COMMENTED** [review]
Small comments

**Comment by sebalix** [comment]
You also have to add an `oca_dependencies.txt` to add `OCA/account-invoicing` repository with https://github.com/OCA/account-invoicing/pull/682

File: base_ebill_payment_contract/views/ebill_payment_contract.xml [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+<!-- Copyright 2020 Camptocamp SA
+     License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl). -->
+<odoo>
+    <record model="ir.ui.view" id="ebill_payment_contract_form_view">
+        <field name="name">ebill.payment.contract.form</field>
+        <field name="model">ebill.payment.contract</field>
+        <field name="arch" type="xml">
+            <form>
+                <sheet>
+                    <group>
```
Comment (sebalix) [inline base_ebill_payment_contract/views/ebill_payment_contract.xml]:
```suggestion
                    <group name="base">
```
(or another name, I'm not inspired :p)

**Review by sebalix — APPROVED** [review]

File: base_ebill_payment_contract/models/res_partner.py [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+class ResPartner(models.Model):
+
+    _inherit = "res.partner"
+
+    def get_active_contract(self, transmit_method):
+        """Return the active contract for a specific transmit method."""
+        self.ensure_one()
+        # FIXME: Search on non stored field
+        contracts = self.env["ebill.payment.contract"].search(
+            [("is_valid", "=", True), ("partner_id", "=", self.id)], limit=1,
```
Comment (guewen) [inline base_ebill_payment_contract/models/res_partner.py:19]: We have `Non-stored field ebill.payment.contract.is_valid cannot be searched.` in logs.

2 options:

* Implement a "search" method on the computed field
* Search all the contracts for the partner (remove limit) and filter them afterwards, it seems acceptable as the number of contract per partner will be a few. 

Solution 2 may be easier, but for UX, being able to search on the `is_valid` field may be useful if we have a list view with all the contracts.

File: base_ebill_payment_contract/__manifest__.py [hunk 7]
```diff
@@ -0,0 +1,15 @@
+# Copyright 2020 Camptocamp SA
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+{
+    "name": "Base Ebill Payment Contract",
```
Comment (yvaucher) [inline base_ebill_payment_contract/__manifest__.py]:
```suggestion
    "name": "Base eBill Payment Contract",
```

File: base_ebill_payment_contract/__manifest__.py [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+    "summary": """
+        Base for ebilling addons""",
```
Comment (yvaucher) [inline base_ebill_payment_contract/__manifest__.py]:
```suggestion
        Base for e-billing addons""",
```

File: base_ebill_payment_contract/models/ebill_payment_contract.py [hunk 9]
```diff
@@ -0,0 +1,50 @@
+# Copyright 2020 Camptocamp SA
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import api, fields, models
+
+
+class EbillPaymentContract(models.Model):
+
+    _name = "ebill.payment.contract"
+    _description = "Ebill Payment Contract"
```
Comment (yvaucher) [inline base_ebill_payment_contract/models/ebill_payment_contract.py]:
```suggestion
    _description = "eBill Payment Contract"
```

File: base_ebill_payment_contract/views/ebill_payment_contract.xml [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+    </record>
+    <record model="ir.actions.act_window" id="ebill_payment_contract_act_window">
+        <field name="name">Ebill Payment Contract</field>
+        <field name="res_model">ebill.payment.contract</field>
+        <field name="view_mode">tree,form</field>
+        <field name="domain">[]</field>
+        <field name="context">{}</field>
+    </record>
+    <record model="ir.ui.menu" id="ebill_payment_contract_menu">
+        <field name="name">Ebill Payment Contract</field>
```
Comment (yvaucher) [inline base_ebill_payment_contract/views/ebill_payment_contract.xml]:
```suggestion
        <field name="name">eBill Payment Contract</field>
```

**Review by yvaucher — CHANGES_REQUESTED** [review]
just a review about terminology.

Reference: https://www.ebill.ch/en/home.html

File: base_ebill_payment_contract/views/ebill_payment_contract.xml (hunk 4 above)
Comment (TDu) [reply base_ebill_payment_contract/views/ebill_payment_contract.xml]: Has been changed.

File: base_ebill_payment_contract/models/res_partner.py (hunk 6 above)
Comment (TDu) [reply base_ebill_payment_contract/models/res_partner.py:19]: I implemented the search method on the third commit.

File: base_ebill_payment_contract/views/ebill_payment_contract.xml (hunk 10 above)
Comment (TDu) [reply base_ebill_payment_contract/views/ebill_payment_contract.xml]: Thanks for the search on terminology, I fixed them all

Comment (yvaucher) [reply base_ebill_payment_contract/views/ebill_payment_contract.xml]: Seems this line slipped out of the fixing

eBill is still written with a lowercase at the start of a sentence.

An other orthography in "E-bill" with the dash. But most people have seen it as "eBill".

File: base_ebill_payment_contract/views/ebill_payment_contract.xml [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+                <field name="partner_id" />
+                <field name="state" />
+                <field name="date_start" />
+                <field name="date_end" />
+                <field name="is_valid" invisible="1" />
+            </tree>
+        </field>
+    </record>
+    <record model="ir.actions.act_window" id="ebill_payment_contract_act_window">
+        <field name="name">Ebill Payment Contract</field>
```
Comment (yvaucher) [inline base_ebill_payment_contract/views/ebill_payment_contract.xml]:
```suggestion
        <field name="name">eBill Payment Contract</field>
```

File: base_ebill_payment_contract/models/res_partner.py [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+
+
+class ResPartner(models.Model):
+
+    _inherit = "res.partner"
+
+    def get_active_contract(self, transmit_method):
+        """Return the active contract for a specific transmit method."""
+        self.ensure_one()
+        # FIXME: Search on non stored field
```
Comment (yvaucher) [inline base_ebill_payment_contract/models/res_partner.py]: You can remove this FIXME if it was for `is_valid`

File: base_ebill_payment_contract/models/res_partner.py [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+        contracts = self.env["ebill.payment.contract"].search(
+            [("is_valid", "=", True), ("partner_id", "=", self.id)], limit=1,
+        )
+        if not contracts:
+            _logger.error(
+                "Paynet contract for {} on {} not found".format(
+                    self.name, transmit_method.name
+                )
+            )
+        return contracts[0] if contracts else contracts
```
Comment (yvaucher) [inline base_ebill_payment_contract/models/res_partner.py:27]: if you name your variable `contract` without the `s` doesn't it make it more obvious you are doing a search with `limit=1` ?

**Review by yvaucher — APPROVED** [review]
Other than the 2 remaining "Ebill" terms to change to "eBill" and a couple of nitpicking

 LGTM

**Review by jgrandguillaume — APPROVED** [review]
Functional test ok

File: base_ebill_payment_contract/models/res_partner.py (hunk 12 above)
Comment (TDu) [reply base_ebill_payment_contract/models/res_partner.py]: Done

**Pull request merged**